Normative language improvements (#158)
AMWA-TV/nmos-discovery-registration: docs/4.2. Behaviour - Querying.md

`authorization` Indicates whether the WebSocket connection requires authorization in order to connect. Use of authorization is likely to be a deployment decision and be the same for all subscriptions exposed from a single Query API, and for the RESTful API itself. Query API clients MAY specify a requested value for this attribute, but in most circumstances this will result in a 400 response code if the Query API is operating in the opposite mode.

## Creating a WebSocket subscription

In order to connect to a WebSocket, a client MUST first request a subscription of a particular type and with particular query parameters by performing a POST as defined in the [Query API](../APIs/QueryAPI.raml) subscriptions documentation. Use of GET requests to find existing suitable WebSocket connections is strongly discouraged.

There is no mandated URL base path for servers to use when creating WebSocket connections. Instead clients SHOULD observe the value of 'ws_href' which is returned by their subscription request in order to identify what to connect to.

WebSockets created for specific clients will be cleaned up automatically if they disconnect.

> Connection upgrade procedures permitting an existing HTTP query to transition to a WebSocket are not currently supported, but could be in the future.

Reviewer: Maybe change "supported" to indicate "covered by this specification" better? Although, see next section, first para!

### Query API Behaviour

> Long-lived connections to the Query API are supported via WebSockets which can be set up via the /subscriptions resource. Query APIs MAY additionally support the HTTP 'Upgrade' header sent by clients to upgrade an HTTP GET request to a WebSocket. In cases where this is performed, a corresponding entry in /subscriptions MUST also be created with matching query parameters.

Reviewer: Hmm, I thought we'd dealt with the fact that the sentence about HTTP 'Upgrade' here seems to cover precisely the thing that the previous section isn't currently supported... Need to check previous discussion.

Reviewer: It was #149. Done.

> Upon receiving a request for a new subscription, the Query API SHOULD identify whether it already has any WebSockets open which provide for the requested query. If a WebSocket exists, the 'subscription' object representing it MAY be returned to the user. If a relevant WebSocket does not exist a new one SHOULD be created.

Reviewer: What does having a WebSocket "open" mean? Not a connection from a client, just whether or not it's currently listening for connections. Would be clearer if it just talked about having an appropriate subscription object...

Author: Would be better resolved through #162

> Upon receiving a request for a new subscription, the Query API SHOULD identify whether it already has any WebSockets open which provide for the requested query. If a WebSocket exists, the 'subscription' object representing it MAY be returned to the user. If a relevant WebSocket does not exist a new one SHOULD be created.

Author: ```suggestion
Upon receiving a request for a new subscription, the Query API SHOULD identify whether it already has a `subscription` object which provides for the requested query, and if so MAY return the `subscription` object to the user. If a relevant `subscription` object does not exist a new one SHOULD be created.
```

Author: Would be better resolved through #162

> In normal operation, once a WebSocket has no more clients subscribed to it the Query API MAY automatically remove it and its corresponding HTTP-advertised subscription object. The Query API MUST NOT acknowledge HTTP DELETE requests for WebSockets running in this 'non-persistent' mode, instead issuing an HTTP 403 response.

Author: ```suggestion
The Query API MAY remove any `subscription` objects that no longer have clients. The Query API MUST NOT acknowledge HTTP `DELETE` requests for WebSockets running in this 'non-persistent' mode, instead issuing an HTTP `403` (Forbidden) response.
```

Author: Would be better resolved through #162

> If a 'persistent' WebSocket has been requested by the client, this MUST NOT be cleaned up automatically by the Query API, even if all consuming clients have disconnected. The user MAY request closing of this socket by issuing an HTTP DELETE. If an HTTP DELETE is issued prior to all WebSocket connections being closed, they SHOULD be forcibly closed by the server.

Reviewer: I have the same slight confusion here about what "a WebSocket" is. I'd rewrite those paragraphs with WebSocket replaced by subscription (and e.g. change 'running' to 'configured').

Author: Will look at these. Agree that talking about subscriptions is more appropriate.

Author: Added suggested versions for two paras, but I'm not sure about the third one. I think this comes down to what 'persistent' can apply to. @andrewbonney any comments?

Author: Would be better resolved through #162
